cdap-tms/src/main/java/co/cask/cdap/messaging/subscriber/AbstractMessagingPollingService.java

```diff
@@ -0,0 +1,267 @@
+    super(retryStrategy);
+    this.topicId = topicId;
+    this.metricsContext = metricsContext;
+    this.fetchSize = fetchSize;
+    this.emptyFetchDelayMillis = emptyFetchDelayMillis;
+  }
+
+  /**
+   * Returns the {@link TopicId} that this service is fetching from.
+   */
+  protected final TopicId getTopicId() {
+    return topicId;
+  }
+
+  /**
+   * Returns the {@link MessageContext} that this service used for interacting with TMS.
+   */
+  protected abstract MessagingContext getMessagingContext();
+
+  /**
+   * Decodes the raw {@link Message} into an object of type {@code T}.
+   *
+   * @param message the {@link Message} to decode
+   * @return an object of type {@code T}
+   * @throws Exception if the decode failed and the given message will be skipped for processing
+   */
+  protected abstract T decodeMessage(Message message) throws Exception;
+
+  /**
+   * Loads last persisted message id.
+   * The returned message id will be used as the starting message id (exclusive) for the first fetch.
+   *
+   * @return the last persisted message id or {@code null} to have first fetch starts from the first available message
+   *         in the topic.
+   * @throws IOException if failed to load the message id
+   */
+  @Nullable
+  protected abstract String loadMessageId() throws IOException;
+
+  /**
+   * Processes the give list of messages. If {@link Exception} is raised from this method,
+   * all messages as provided through the {@code messages} parameter will be replayed in the next call.
+   *
+   * @param messages an {@link Iterator} of {@link ImmutablePair}, with the {@link ImmutablePair#first}
+   *                 as the message id, and the {@link ImmutablePair#second} as the decoded message
+   * @return the message ID for the next fetch to start from (exclusively). If returning {@code null},
+   *         the next fetch will start from the same message ID that was used to fetch the current batch of
+   *         messages
+   * @throws Exception if there is error processing messages.
+   */
+  @Nullable
+  protected abstract String processMessages(Iterator<ImmutablePair<String, T>> messages) throws Exception;
+
+  /**
+   * Perform post processing after a batch of messages has been processed and before the next batch of
+   * messages is fetched.
+   */
+  protected void postProcess() {
+    // no-op
+  }
+
+  @Override
+  protected final boolean shouldRetry(Exception ex) {
+    // Log the exception
+    try {
+      throw ex;
+    } catch (ServiceUnavailableException e) {
+      SAMPLING_LOG.warn("Failed to contact service {}. Will retry in next run.", e.getServiceName(), e);
+    } catch (TopicNotFoundException e) {
+      SAMPLING_LOG.warn("Failed to fetch from TMS. Will retry in next run.", e);
+    } catch (Exception e) {
+      SAMPLING_LOG.warn("Failed to get and process notifications. Will retry in next run", e);
+    }
+
+    return true;
```

Comment:
This is for the `RetryStrategy` to determine if it should retry on a given exception. The event loop termination is determined by the `RetryStrategy` that passed to super class, instead of by this method.

I am also removing the `final` to make this behavior overridable.